File: vane-core/src/main/java/org/oddlama/vane/core/command/params/FixedParam.java
package org.oddlama.vane.core.command.params;

import java.util.Collections;
import java.util.List;
import org.bukkit.command.CommandSender;
import org.oddlama.vane.core.command.Command;
import org.oddlama.vane.core.command.check.CheckResult;
import org.oddlama.vane.core.command.check.ErrorCheckResult;
import org.oddlama.vane.core.command.check.ParseCheckResult;
import org.oddlama.vane.core.functional.Function1;

public class FixedParam<T> extends BaseParam {

    private T fixedArg;
    private String fixedArgStr;
    private boolean includeParam = false;
    private boolean ignoreCase = false;

    public FixedParam(Command<?> command, T fixedArg, Function1<T, String> toString) {
        super(command);
        this.fixedArg = fixedArg;
        this.fixedArgStr = toString.apply(fixedArg);
    }

    /** Will ignore the case of the given argument when matching */
    public FixedParam<T> ignoreCase() {
        this.ignoreCase = true;
        return this;
    }

    /** Will pass this fixed parameter as an argument to the executed function */
    public FixedParam<T> includeParam() {
        this.includeParam = true;
        return this;
    }

    @Override
    public CheckResult checkParse(CommandSender sender, String[] args, int offset) {
        if (args.length <= offset) {
            return new ErrorCheckResult(offset, "§6missing argument: §3" + fixedArgStr + "§r");
        }
        var parsed = parse(args[offset]);
        if (parsed == null) {
            return new ErrorCheckResult(
                offset,
                "§6invalid argument: expected §3" + fixedArgStr + "§6 got §b" + args[offset] + "§r"
            );
        }
        return new ParseCheckResult(offset, fixedArgStr, parsed, includeParam);
    }

    @Override
    public List<String> completionsFor(CommandSender sender, String[] args, int offset) {
        return Collections.singletonList(fixedArgStr);
    }

    private T parse(String arg) {
        if (ignoreCase) {
            if (arg.equalsIgnoreCase(fixedArgStr)) {
                return fixedArg;
            }
        } else {
            if (arg.equals(fixedArgStr)) {
                return fixedArg;
            }
        }

        return null;
    }
}

File: vane-core/src/main/kotlin/org/oddlama/vane/core/command/params/FixedParam.kt
package org.oddlama.vane.core.command.params

import org.bukkit.command.CommandSender
import org.oddlama.vane.core.command.Command
import org.oddlama.vane.core.command.check.CheckResult
import org.oddlama.vane.core.command.check.ErrorCheckResult
import org.oddlama.vane.core.command.check.ParseCheckResult
import org.oddlama.vane.core.functional.Function1

/**
 * Parameter that matches a single fixed literal value.
 *
 * @param T fixed value type.
 * @param command the owning command.
 * @param fixedArg fixed value matched by this parameter.
 * @param toString formatter for the fixed value.
 */
class FixedParam<T>(command: Command<*>?, private val fixedArg: T?, toString: Function1<T?, String?>) :
    BaseParam(command) {
    /**
     * String representation of [fixedArg] used for parsing and completion.
     */
    private val fixedArgStr: String? = toString.apply(fixedArg)

    /**
     * Whether the fixed parameter should be forwarded to executors.
     */
    private var includeParam = false

    /**
     * Whether matching is case-insensitive.
     */
    private var ignoreCase = false

    /**
     * Enables case-insensitive matching.
     */
    fun ignoreCase(): FixedParam<T> = apply { ignoreCase = true }

    /**
     * Includes this matched fixed value in executor arguments.
     */
    fun includeParam(): FixedParam<T> = apply { includeParam = true }

    /**
     * Parses the fixed literal argument.
     */
    override fun checkParse(sender: CommandSender?, args: Array<String?>?, offset: Int): CheckResult {
        if (args == null || args.size <= offset)
            return ErrorCheckResult(offset, "§6missing argument: §3$fixedArgStr§r")
        val argVal = args[offset]
            ?: return ErrorCheckResult(offset, "§6invalid argument: expected §3$fixedArgStr§6 got §bnull§r")
        return parse(argVal)
            ?.let { ParseCheckResult(offset, fixedArgStr ?: "", it, includeParam) }
            ?: ErrorCheckResult(offset, "§6invalid argument: expected §3$fixedArgStr§6 got §b$argVal§r")
    }

    /**
     * Returns the fixed literal as the only completion candidate.
     */
    override fun completionsFor(sender: CommandSender?, args: Array<String?>?, offset: Int): MutableList<String?> =
        mutableListOf(fixedArgStr)

    /**
     * Matches a raw argument against [fixedArgStr].
     */
    private fun parse(arg: String): T? =
        fixedArg.takeIf { arg.equals(fixedArgStr, ignoreCase = ignoreCase) }
}
